concordium-contracts-common/src/schema.rs

```diff
@@ -700,53 +704,52 @@ impl Deserial for FunctionV1 {
 
 impl Serial for FunctionV2 {
     fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
-        match self {
-            FunctionV2::Param(parameter) => {
+        let parameter = self.parameter.is_some();
+        let return_value = self.return_value.is_some();
+        let error = self.error.is_some();
+        match (parameter, return_value, error) {
+            // parameter
+            (true, false, false) => {
                 out.write_u8(0)?;
-                parameter.serial(out)
+                self.parameter.serial(out)
             }
-            FunctionV2::Rv(return_value) => {
+            // return_value
+            (false, true, false) => {
                 out.write_u8(1)?;
-                return_value.serial(out)
+                self.return_value.serial(out)
             }
-            FunctionV2::ParamRv {
-                parameter,
-                return_value,
-            } => {
+            // parameter + return_value
+            (true, true, false) => {
                 out.write_u8(2)?;
-                parameter.serial(out)?;
-                return_value.serial(out)
+                self.parameter.serial(out)?;
+                self.return_value.serial(out)
             }
-            FunctionV2::Error(error) => {
+            // error
+            (false, false, true) => {
                 out.write_u8(3)?;
-                error.serial(out)
+                self.error.serial(out)
             }
-            FunctionV2::ParamError {
-                parameter,
-                error,
-            } => {
+            // parameter + error
+            (true, false, true) => {
                 out.write_u8(4)?;
-                parameter.serial(out)?;
-                error.serial(out)
+                self.parameter.serial(out)?;
+                self.error.serial(out)
             }
-            FunctionV2::RvError {
-                return_value,
-                error,
-            } => {
+            // return_value + error
+            (false, true, true) => {
                 out.write_u8(5)?;
-                return_value.serial(out)?;
-                error.serial(out)
+                self.return_value.serial(out)?;
+                self.error.serial(out)
             }
-            FunctionV2::ParamRvError {
-                parameter,
-                return_value,
-                error,
-            } => {
+            // parameter + return_value + error
+            (true, true, true) => {
                 out.write_u8(6)?;
-                parameter.serial(out)?;
-                return_value.serial(out)?;
-                error.serial(out)
+                self.parameter.serial(out)?;
+                self.return_value.serial(out)?;
+                self.error.serial(out)
             }
+            // no schema: cannot happen
+            (false, false, false) => out.write_u8(7),
         }
     }
 }
```

Comment:
Are you sure this is correct. You seem to be serializing `Option<Type>` all the time, which as far as I understand is not the same as it used to be.

I think what you want is
```rust

impl Serial for FunctionV2 {
    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
        let parameter = self.parameter.is_some();
        let return_value = self.return_value.is_some();
        let error = self.error.is_some();
        let tag: u8 = match (parameter, return_value, error) {
            // parameter
            (true, false, false) => 0,
            // return_value
            (false, true, false) => 1,
            // parameter + return_value
            (true, true, false) => 2,
            // error
            (false, false, true) => 3,
            // parameter + error
            (true, false, true) => 4,
            // return_value + error
            (false, true, true) => 5,
            // parameter + return_value + error
            (true, true, true) => 6,
            // no schema: cannot happen
            (false, false, false) => 7,
        };
        out.write_u8(tag)?;
        if let Some(p) = self.parameter.as_ref() {
            p.serial(out)?;
        }
        if let Some(rv) = self.return_value.as_ref() {
            rv.serial(out)?;
        }
        if let Some(err) = self.error.as_ref() {
            err.serial(out)?;
        }
        Ok(())
    }
}
```

Or am I misunderstanding something?